service/controller/azuremachinepool/handler/nodepool/create_wait_for_old_workers_to_be_gone.go

```diff
@@ -0,0 +1,97 @@
+package nodepool
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-07-01/compute"
+	"github.com/giantswarm/microerror"
+	"sigs.k8s.io/cluster-api/util"
+
+	"github.com/giantswarm/azure-operator/v5/pkg/handler/nodes/state"
+	"github.com/giantswarm/azure-operator/v5/pkg/tenantcluster"
+	"github.com/giantswarm/azure-operator/v5/service/controller/key"
+)
+
+func (r *Resource) waitForOldWorkersToBeGoneTransition(ctx context.Context, obj interface{}, currentState state.State) (state.State, error) {
+	azureMachinePool, err := key.ToAzureMachinePool(obj)
+	if err != nil {
+		return DeploymentUninitialized, microerror.Mask(err)
+	}
+
+	cluster, err := util.GetClusterFromMetadata(ctx, r.CtrlClient, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return DeploymentUninitialized, microerror.Mask(err)
+	}
+
+	if !cluster.GetDeletionTimestamp().IsZero() {
+		r.Logger.Debugf(ctx, "Cluster is being deleted, skipping reconciling node pool")
+		return currentState, nil
+	}
+
+	tenantClusterK8sClient, err := r.tenantClientFactory.GetClient(ctx, cluster)
+	if tenantcluster.IsAPINotAvailableError(err) {
+		r.Logger.Debugf(ctx, "tenant API not available yet")
+		r.Logger.Debugf(ctx, "canceling resource")
+
+		return currentState, nil
+	} else if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	virtualMachineScaleSetVMsClient, err := r.ClientFactory.GetVirtualMachineScaleSetVMsClient(ctx, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	virtualMachineScaleSetsClient, err := r.ClientFactory.GetVirtualMachineScaleSetsClient(ctx, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	r.Logger.Debugf(ctx, "checking if there are old nodes still running")
+	var allWorkerInstances []compute.VirtualMachineScaleSetVM
+	{
+		r.Logger.Debugf(ctx, "finding all worker VMSS instances")
+
+		allWorkerInstances, err = r.GetVMSSInstances(ctx, virtualMachineScaleSetVMsClient, key.ClusterID(&azureMachinePool), key.NodePoolVMSSName(&azureMachinePool))
+		if err != nil {
+			return DeploymentUninitialized, microerror.Mask(err)
+		}
+
+		r.Logger.Debugf(ctx, "found %d worker VMSS instances", len(allWorkerInstances))
+	}
+
+	resourceGroupName := key.ClusterID(&azureMachinePool)
+	nodePoolVMSSName := key.NodePoolVMSSName(&azureMachinePool)
+	vmss, err := virtualMachineScaleSetsClient.Get(ctx, resourceGroupName, nodePoolVMSSName)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	var oldWorkersCount int
+	{
+		for _, i := range allWorkerInstances {
+			old, err := r.isWorkerInstanceFromPreviousRelease(ctx, tenantClusterK8sClient, azureMachinePool.Name, i, vmss)
```

Comment:
I'd suggest to re-design this a bit. Given I read the code correctly, this seems to now fetch VMSS instances, then iterate them and for each instance list WC nodes to find the corresponding pair.

Alternative proposal:

After fetching VMSS instances, `List()` WC worker nodes (w/ label selector) and convert slice into `map[string]corev1.Node` that is keyed by node name. Then you could change this function call into:
```go
n, exists := nodes[key.NodePoolInstanceName(azureMachinePool.Name, *instance.InstanceID)]
if !isWorkerInstanceFromPreviousRelease(n) || !isWorkerInstanceWithLatestSKU(i, vmss) {
  oldWorkersCount += 1
}
```

Similar approach then for `sortNodesByTenantVMState()` below.

Reply:
Thanks for the suggestion, I didn't like that part either and was about to refactor it today.
I am actually thinking of closing this PR and open a new one once finished the refactoring, but don't want you to be upset because you did the review already. WDYT?

Reply:
I'm totally fine both ways. Whatever works better for you. :+1: